karina-tracy/scripts/article.js

```diff
@@ -0,0 +1,96 @@
+Article.prototype.toHtml = function() {
+  let template = Handlebars.compile($('#article-template').text());
+
+  this.daysAgo = parseInt((new Date() - new Date(this.publishedOn))/60/60/24/1000);
+
+  // COMMENT: What is going on in the line below? What do the question mark and colon represent? How have we seen this same logic represented previously?
+  // Not sure? Check the docs!
+  // The question mark and colon are ternary operator. This works as the if, else statements. The question mark works as the if and the colon would be the else. So if something is true (?) then do this (:) if its not do this.
+  this.publishStatus = this.publishedOn ? `published ${this.daysAgo} days ago` : '(draft)';
+  this.body = marked(this.body);
+
+  return template(this);
+};
+
+// REVIEW: There are some other functions that also relate to all articles across the board, rather than just single instances. Object-oriented programming would call these "class-level" functions, that are relevant to the entire "class" of objects that are Articles.
+
+// REVIEW: This function will take the rawData, how ever it is provided, and use it to instantiate all the articles. This code is moved from elsewhere, and encapsulated in a simply-named function for clarity.
+
+// COMMENT: Where is this function called? What does 'rawData' represent now? How is this different from previous labs?
+// .loadAll() is called in the .fetchAll() function. Before .fetchAll() function, rawData was an array of objects (our main data). In .fetchAll() function, rawData becomes a value of the local storage. .fetchAll() is called from .getEtag which is called from index.html after the page loads.
+Article.loadAll = articleData => {
+  articleData.sort((a,b) => (new Date(b.publishedOn)) - (new Date(a.publishedOn)))
+
+  articleData.forEach(articleObject => Article.all.push(new Article(articleObject)))
+}
+
+// REVIEW: This function will retrieve the data from either a local or remote source, and process it, then hand off control to the View.
+Article.fetchAll = () => {
+  // REVIEW: What is this 'if' statement checking for? Where was the rawData set to local storage?
+  //The if statement is checking is the local storage has data. rawData was set to local storage in the callback function passed to $.getJSON.
+
+  if (localStorage.rawData) {
+    // If localStorage.rawData exists then use the cashed data
+    Article.loadAll(JSON.parse(localStorage.rawData));
+    articleView.initIndexPage();
+  } else {
+    // Otherwise, kick off two asych functions:
+    //    $.getJSON to get new data from the data source
+    $.getJSON('data/hackerIpsum.json')
+      .then(data => {
+        localStorage.setItem('rawData', JSON.stringify(data));
+        Article.loadAll(data)
+        articleView.initIndexPage();
+      })
+      .catch(err => {
+        console.error('Error on fetching data/hackerIpsum.json',err);
+      });
+    //   and $.ajax to get and save the new eTag value
```

Comment:
If you're using `.ajax` to make a request to that file, you don't need the `getJSON` call above! Since this is just a call to store the `eTag` header, and that'll be sent for any request, you can change this to a `GET` request, and save both the data and the `eTag` header on success.